### ISSUE TITLE
[models] Ensure all PyTorch models are ONNX exportable

### ISSUE THREAD

**Issue by fg-mindee** [mindee/doctr#789]
Most users of the library are more interested in existing pretrained models to use for inference rather than training. For this reason, it's important to ensure we can easily export those trained models.

- General
    - [x] Add a unittest/CI job to check that all models can be exported with ONNX (#976, #979, #980 )

- Classification
     - [x] VGG (#830)
    - [x] ResNet (#830)
    - [x] MobileNet V3 (#830)
    - [x] MAGC ResNet (#830)

- Text Detection
    - [x] LinkNet #979 
    - [x] DBNet #979 

- Text Recognition
    - [x] CRNN #980 
    - [x] SAR #980 
    - [x] MASTER #980  (#986)
    
    NOTE: master is exportable but in fact of an onnxruntime internal issue (https://github.com/microsoft/onnxruntime/issues/10994) the loading fails

**Comment by felixdittrich92** [comment]
@charlesmindee  classification complete :+1:

**Comment by frytoli** [comment]
I'm relatively green when it comes to ML engineering, but I've been attempting to look at exporting the DBNet models to ONNX. I think we're relatively close already and just need to convert inputs/outputs to a data structure compatible with ONNX (specifically not numpy.ndarray). Here's some sample code and the resulting error. I'm slowly working on solving this on my end, but I wanted to post it here in case this is a really easy problem to solve for someone else more familiar with the project.

```
import torch
from doctr.models.detection import db_resnet50_rotation
from doctr.models.utils import export_classification_model_to_onnx

model = db_resnet50_rotation(pretrained=True)
torch.onnx.export(model, torch.randn(1, 3, 1024, 1024), "model.onnx", input_names=['input'], output_names=['output'], opset_version=13)
```

```
RuntimeError: Only tuples, lists and Variables are supported as JIT inputs/outputs. Dictionaries and strings are also accepted, but their usage is not recommended. Here, received an input of unsupported type: numpy.ndarray
```

**Comment by felixdittrich92** [comment]
Hi @frytoli :wave:,

thanks for working on this :hugs: 

To clearify where the problem is: 
If you run the following with `verbose=True` you get a more detailed output:

```
import torch
from doctr.models.detection import db_resnet50_rotation

model = db_resnet50_rotation(pretrained=True)
dummy_input = torch.rand((1, 3, 1024, 1024), dtype=torch.float32)
torch.onnx.export(
        model,
        dummy_input,
        "test.onnx",
        input_names=['input'],
        output_names=['logits'],
        dynamic_axes={'input': {0: 'batch_size'}, 'logits': {0: 'batch_size'}},
        export_params=True, opset_version=13, verbose=True
    )
```

problem: `preds[0] for preds in self.postprocessor(prob_map.detach().cpu().permute((0, 2, 3, 1)).numpy())`

Currently the models forward pass includes the postprocessing call. 
What in my opinion is not optimal to avoid something like this, when modeling, it should be noted that the model only returns specific data, i.e. logits vector and, if ground truth is present, loss. 
So what should be done in the first place is to make these changes for all detection and recognition models. 
However, doing so would break the pure pipeline, so we would have to see how it affects runtime/latency. 
Furthermore, we would have to extend these changes to the Tensorflow part as well, in order to maintain equality on the one hand and on the other hand I think that we can then export the models more easily in Tensorflow's SavedModel format. 
If you are interested in it, please open a pull request and we can work on it together if you like. 
@charlesmindee @frgfm what is your opinion about this ?

**Comment by charlesmindee** [comment]
Hi @felixdittrich92 and @frytoli, 
Thanks for bringing this on the table, to avoid breaking the wore pipeline as Felix mentioned it, we may want to create a flag ("exportable" or "raw" or something like that) that could be passed in the model initialization: `db_resnet50(pretrained=True, exportable=True)`, and this would  return in the call of the model the raw output without the postprocessing, or with an adapted post processing (ONNX compatible), what do you think ?
@MaximeChurin

**Comment by felixdittrich92** [comment]
@charlesmindee  @frytoli  Sounds good to me, but we have to take care of increasing complexity and the differences between TF and PT implementations. So i would suggest maybe to try it with on model but on both implementations so that we can provide an equivalent implementation which solves the ONNX export problem on PT and the SavedModel export on TF side. Wdyt ?

**Comment by felixdittrich92** [comment]
So maybe if we can handle both with Onnx do we really need TF's SavedModel also ?

**Comment by frgfm** [comment]
Hi all 👋 

So, yes, we're very aware of the JIT incompatibility with some data structures & numpy operations. Having conditional executions in call methods could become tricky for maintenance later on

This is a big topic, hence this issue. Here is a suggestion:
- all models's "core computation" should be compatible with JIT export
- we separate the post-processing 
- using ONNX export, the post-processing will still need to be put afterwards.

What do you think?

**Comment by felixdittrich92** [comment]
@frgfm @charlesmindee 
Hi again :wave: ,
I see two options now:

1. We make each model exportable by just going to the logits vector and do post-processing afterwards. @frgfm 
2. We try to make the post-processing step JIT compatible. @charlesmindee 

Keep in mind as a user i think the most than want also to use this exported models inside docTR.
So maybe a bit offtopic and not directly matching with this issue but why not providing the training part of this library in PT and TF and the inference pipeline with onnxruntime some points i see:
1. It would be much easier to maintain
2. If we really have problems with a model in a specific framework than if you want you could use the working one with a pure onnx inference pipeline no problem
3. Now with the next release we provide a model sharing option so we have to provide that it is backward compatible (i am really a bit scared about this part :sweat_smile: ) would be no problem with onnx 
4. we could more focus on optimization for the inference and have a clear split between training and inference

I know this would be a lot of refactoring but maybe it could be a good match

What do you think ? :smile: 
I think at this point we have to find a good way which will lead to breaking changes before we can go ahead !

**Comment by frgfm** [comment]
My bad, I should clarify my previous thought:
- if we can make all the steps JIT compatible, yes sure, we should do it. In my previous attempts, I remember that this would require diving into obscure cv2 functions (which aren't compatible as is)
- if it's not possible or quite hard, I suggest preserving a post-processing after the ONNX part

About the training & alignment with TF, this is a serious problem. Moving between TF & ONNX is much more troublesome than between PyTorch & ONNX. I fully agree that we should do our best to go into the right direction though 😅

**Comment by felixdittrich92** [comment]
short update:
both SAR implementations are exportable with ONNX :100:   (only a small flaw for TF we have to provide a constant batch size. Why ? If we pass the dummys batch size as None than it is not possible to compute the initial LSTMCell states)

left:
classification: MAGC (only TF)
recognition: MASTER (TF and PT)

**Comment by felixdittrich92** [comment]
update: with #941 PyTorch is done :heart: 
All models can be exported only MAGC and MASTER on TF left

One ref: https://github.com/microsoft/onnxruntime/issues/10994 MASTER (PT) model is exportable but doesn't work with onnxruntime currently until this ticket is closed

**Comment by felixdittrich92** [comment]
@frgfm @charlesmindee I think this issue is only about the export i would open another issue for the loading part wdyt ? :)

**Comment by frgfm** [comment]
@felixdittrich92 yes precisely :+1:

**Comment by felixdittrich92** [comment]
@frgfm @charlesmindee I think it is ok if we close this all models (without postprocessing) can be exported
